### DISCUSSION TITLE
[Question][Hw5] Traverse the graph

### DISCUSSION THREAD

**Discussion by john-mai-2605** [prosyslab-classroom/cs348-information-security#621] [Q&A]
How could I traverse the graph to update the table. Seems like there is no function to get the next node (only previous by pred)

**Comment by RiceBiscuits** [comment]
You should implement `run` function in Analysis.ml to traverse the graph and find a fixed point.

**Comment by john-mai-2605** [comment]
Do you mean to traverse the graph by the table keys?

**Comment by RiceBiscuits** [comment]
`let run llctx table = failwith "Not implemented"
`
A table and an llcontext will be provided in the run function, so you can use a table.

**Comment by john-mai-2605** [comment]
Thank you for your time.
Yes, I understand I have to modify the run.
My question is, how to traverse the graph when we have no function to get the next node
Is it right to just traverse the keys (keys of the table are nodes) of the table?

**Comment by mickowale** [comment]
Since you can't get the next node (with the skeleton code), try to cover every single edge in the graph. Forget actually traversing the graph and focus on each edge. For this, the `pred ` function should be enough. The goal is, by doing this enough times, you should get a fixed point.
